enhancements/cluster-logging/cluster-logging-log-forwarding.md

```diff
@@ -0,0 +1,310 @@
+
+Organizations desire to reuse their existing enterprise log solutions to store container logs.  Providing a declarative mechanism by which administrators define a log destination simplifies their operational burden.  The cluster logging stack is able to deploy a collector to each OKD node that is configured with the necessary permissions to collect the logs, add container metadata (e.g. labels) and ship them to the specified endpoint.
+
+
+
+### Goals
+The specific goals of this proposal are:
+
+* Provide an API that identifies log sources and their desired destination
+* Send logs to an Elasticsearch cluster not managed by the cluster logging infrastructure
+* Send logs to a fluentd instance not managed by the cluster logging infrastructure via fluentd's secureforward protocol
+* Allow deployment of log forwarding without deploying the entirety of the cluster logging infrastructure (e.g. Kibana, Elasticsearch)
+* Support TLS between the collector and destination if so configured
+* Specifically only forward "application" and "infrastructure" logs
+* Re-use log forwarding to write logs to a cluster logging managed log aggregator (e.g. Elasticsearch)
+
+We will be successful when:
+
+* an administrator is able to deploy their own log aggregation service
+* specifies this service in log forwarding configuration
+* the service receives logs from a OKD cluster logging deployed collection agent  
+
+
+
+### Non-Goals
+
+Log forward intends to be opinionated about the source of log messages; it will explicitly define what constitutes "application logs" or "infrastructure logs". It will not allow configuration of additional sources without further design considerations and how those sources may impact the collecting agent. Log forwarding is not intended to provide a complex routing solution as one might achieve by using a custom collector configuration or a messaging solution (e.g. kafka).
+
+## Proposal
+
+Log forwarding will provide a declarative way to specify the destination for specific types of logs using a 'pipeline'.  A 'pipeline' defines simple routing for one source to one or more destinations. The source of logs are opinionated and well defined by cluster logging. The initial source types are as follows:
+
+* `logs.app` - Container logs generated by user applications running on the platform, excluding infrastructure container applications
+* `logs.infra` - Logs generated by both infrastructure components running on the platform and OKD nodes (e.g. journal logs).  "Infra" applications are defined as any pods which run in namespaces: `openshift*`, `kube*`, `default`.
+
+Future types may be:
+* `events.kubernetes` - Kubernetes events
+* `audit.container-engine` - Container engine audit events
+
+There are no assumptions regarding whether or not an endpoint is deployed on or off cluster.  Endpoints off-cluster may require adminstrators to perform additional actions in order for logs to be forwarded (e.g. secret creation, opening port) 
```

Comment:
may also require proxy configuration, CA injection.

Reply:
So, for example, the steps for configuring secure_forward would be:
* user creates `secret/my-secure-forward-secret` which contains the "secret" parameters (that is, the parameters whose values we don't want in cleartext in the clusterlogging CR) used to configure https://docs.fluentd.org/output/forward:
```yaml
password: ASJDWEASJDFasd==
shared_key: KV203lnJSDFlasdf==
tls_cert_path: ASDJFASDFASDF==
tls_client_cert_path: XCKFWE2308945==
tls_client_private_key_path: ZXMCVOOEPWEP99==
etc.
```
well, the `path` parameters would be the actual contents of the cert/key files - the secret would mount these files into the pod, then we would generate fluentd config like
```
    tls_cert_path /etc/fluent/my-secure-forward-secret/tls_cert_path
```

Reply:
> may also require proxy configuration, CA injection.

Are admins doing anything special in this situation? Would they not create the global proxy object and then operator would consume it without any specific config by an admin for forwarding?

Reply:
they might need to configure their proxy to allow access to the external logstore, or list it in a noproxy setting.

Or if for some reason this is the first/only thing that needs to go through the proxy, then they'd have to setup the cluster proxy config from scratch for this.

i'd put it in the same category as "opening a port"


Also your operator would need to wire the proxyconfig+CAs to fluentd(or to whatever thing needs to forward the logs externally), if it is not already.

Reply:
We would need to make sure our operator looks for the `proxy` config object and if it exists leverage the `httpProxy`, `httpsProxy`, `noProxy`, and `additionalTrustBundle` information.

Reply:
@bparees please advise how to reflect your comment in this proposal

Reply:
1) add proxy configuration to the list of things they may need to do, in addition to opening ports

2) basically what @ewolinetz said.  we need to ensure that when we configure our fluentd, we are providing it with the cluster proxy configuration/CAs so it can use them when forwarding logs off cluster.  I think you guys already took a pass at doing that in 4.2 though it may not be complete?

Reply:
modified to add `enable global proxy configuration`